lib/SILOptimizer/Utils/OwnershipOptUtils.cpp

```diff
@@ -353,20 +368,53 @@ bool OwnershipRAUWHelper::hasValidRAUWOwnership(SILValue oldValue,
 // extend the lifetime of \p oldValue to cover the new uses.
 static bool canFixUpOwnershipForRAUW(SILValue oldValue, SILValue newValue,
                                      OwnershipFixupContext &context) {
-  if (!OwnershipRAUWHelper::hasValidRAUWOwnership(oldValue, newValue))
-    return false;
+  switch (oldValue.getOwnershipKind()) {
+  case OwnershipKind::Guaranteed:
+    // Check that the old lifetime can be extended and record the necessary
+    // book-keeping in the OwnershipFixupContext.
+    context.clear();
+
+    // Check that no transitive uses have a PointerEscape, and record the leaf
+    // uses for liveness extension.
+    if (!findExtendedTransitiveGuaranteedUses(oldValue,
+                                              context.guaranteedUsePoints))
+      return false;
+    return OwnershipRAUWHelper::hasValidRAUWOwnership(
+        oldValue, newValue, context.guaranteedUsePoints);
+  case OwnershipKind::Owned: {
+    SmallVector<Operand *, 8> ownedUsePoints;
+    // If newValue is lexical, find the uses of oldValue so that it can be
+    // determined whether the replacement would illegally extend the lifetime
+    // of newValue.
+    if (newValue->isLexical() &&
+        !findUsesOfSimpleOwnedValue(oldValue, &ownedUsePoints))
+      return false;
+    return OwnershipRAUWHelper::hasValidRAUWOwnership(oldValue, newValue,
+                                                      ownedUsePoints);
+  }
+  default:
+    return OwnershipRAUWHelper::hasValidRAUWOwnership(oldValue, newValue, {});
```

Comment:
> default:
>     return OwnershipRAUWHelper::hasValidRAUWOwnership(oldValue, newValue, {});

I think the same lifetime extension happens when an unowned value's uses is replaced with a guaranteed or owned value. So it should probably be just like the owned case here.

Reply:
Fixed.